Add `round`, `floor`, `ceil` for float tensor

med1844 opened the pull request:
## Pull Request Template

### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
  However, I'm consistently encountering failures when running tests for `burn_wgpu`:
  ```
  [2024-10-14T04:16:15Z INFO  tracel_xtask::utils::process] Unit Tests: burn_wgpu
       Running unittests src/lib.rs (target/debug/deps/burn_wgpu-8e485c42efde75ee)
  
  running 1641 tests
  error: test failed, to rerun pass `-p burn-wgpu --lib`
  
  Caused by:
  [2024-10-14T04:16:16Z INFO  tracel_xtask::utils::process] Unit Tests: burn_wgpu
    process didn't exit successfully: `burn/target/debug/deps/burn_wgpu-8e485c42efde75ee --color=always` (signal: 11, SIGSEGV: invalid memory reference)
  Error: Workspace Unit Tests failed
  ```
  Any suggestions on how to resolve, debug, or bypass this test set would be appreciated.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

#2360

### Changes

I've implemented `round`, `floor`, and `ceil` for multiple backends to the best of my ability. There are a couple of points worth noting:

#### `burn-jit` not supported currently

Implementation for `burn-jit` is pending due to `cubecl` limitations. The issue seems to be that the Round, Floor, and Ceil traits defined in this file:

https://github.com/tracel-ai/cubecl/blob/5fa46c9201671643315b8c47f60bcdfa0566a8e2/crates/cubecl-core/src/frontend/operation/unary.rs#L1

are not implemented for `Line`:

https://github.com/tracel-ai/cubecl/blob/5fa46c9201671643315b8c47f60bcdfa0566a8e2/crates/cubecl-core/src/frontend/container/line/ops.rs

I have added `todo!()` in the corresponding methods in `src/ops/float_ops.rs` as a placeholder.

#### Gradient of step functions

Based on this [discussion](https://discuss.pytorch.org/t/torch-round-gradient/28628) on the PyTorch forum, I've set the gradient to all zeros. While I'm not certain this is the optimal approach, it aligns with PyTorch's implementation. You might want to review the decision made here.

### Testing

I've added tests for burn-tensor, burn-tch, burn-ndarray, and burn-candle. However, I couldn't find existing tests for tensor operations in burn-tch and burn-fusion. Is this expected for these modules?

laggui reviewed (changes requested):
Thanks for adding these!

Overall, the implementation looks good! Just a note regarding rounding which we should probably be careful about.

If we merge this, we should open an issue to add the missing operations for JIT backends or add a default implementation before merging.

med1844 commented on crates/burn-tensor/src/tensor/ops/tensor.rs:
```diff
@@ -931,6 +931,39 @@ pub trait FloatTensorOps<B: Backend> {
     /// A tensor with the same shape as `tensor` with tangent values.
     fn float_tanh(tensor: FloatTensor<B>) -> FloatTensor<B>;
 
+    /// Returns a new tensor with rounded values.
+    ///
+    /// # Arguments
+    ///
+    /// * `tensor` - The tensor to be rounded.
+    ///
+    /// # Returns
+    ///
+    /// A tensor with the same shape as `tensor` with rounded values.
```
Tch do this by default. For NdArray we can use `f64::round_ties_even(x)` as you have mentioned. Candle doesn't seem to provide round to even by default: https://github.com/huggingface/candle/blob/main/candle-core/src/op.rs#L731-L750
I have implemented the tiebreaker strategy but might not be the most efficient solution... maybe we should open an issue there too.

med1844 commented:
> If we merge this, we should open an issue to add the missing operations for JIT backends or add a default implementation before merging.

I took another try and still have no clue how to add default implementation. Any advice on this is appreciated! I have also opened a PR in cubecl (https://github.com/tracel-ai/cubecl/pull/179)

We might also want to test if JIT backend provides round to even strategy.

laggui commented:
@med1844 take a look at the [`float_sort` implementation](https://github.com/tracel-ai/burn/blob/main/crates/burn-tensor/src/tensor/ops/tensor.rs#L1241). By default, the operation calls the [`sort`](https://github.com/tracel-ai/burn/blob/main/crates/burn-tensor/src/tensor/api/sort.rs#L31) function (which, in this case, is executed directly on the tensor data). While this may not be extremely efficient as this could move the data from GPU to CPU, it makes sure that every backend can at least run the operation unless redefined.

Regarding the JIT backends, both CUDA and WGPU runtimes in cubecl define the round operation with round to even so it should be fine.

If you don't have time to add a default implementation or op definition for jit backends we can leave it unimplemented for this PR - as long we track it with an issue 🙂 it's already nice that we have these ops added

med1844 commented:
I have added implementations for JIT backend thanks to cubecl PR being merged, and also for the newly added `burn-router`. From the previous CI error message, it seems that `num_traits` doesn't support `round_tie_even` for `Float` in `no-std` mode, so I added a fallback implementation.

I have passed `run-checks.sh no-std`. But for `run-checks.sh std` I keep getting problems:

```
error: test failed, to rerun pass `-p burn-router --lib`

Caused by:
[2024-10-19T23:38:41Z INFO  tracel_xtask::utils::process] Unit Tests: burn_router
  process didn't exit successfully: `burn/target/debug/deps/burn_router-554570ceab92bef6 --color=always` (signal: 11, SIGSEGV: invalid memory reference)
Error: Workspace Unit Tests failed
```

And no debug information when trying to debug with `rust-gdb`. I could only hope it would pass CI.

laggui reviewed (commented):
Awesome, thanks for adding the implementations! 🎉 

I can take a look at this to try and fix the issue so we can merge this 🙂

laggui commented:
> I have added implementations for JIT backend thanks to cubecl PR being merged, and also for the newly added `burn-router`. From the previous CI error message, it seems that `num_traits` doesn't support `round_tie_even` for `Float` in `no-std` mode, so I added a fallback implementation.
> 
> I have passed `run-checks.sh no-std`. But for `run-checks.sh std` I keep getting problems:
> 
> ```
> error: test failed, to rerun pass `-p burn-router --lib`
> 
> Caused by:
> [2024-10-19T23:38:41Z INFO  tracel_xtask::utils::process] Unit Tests: burn_router
>   process didn't exit successfully: `burn/target/debug/deps/burn_router-554570ceab92bef6 --color=always` (signal: 11, SIGSEGV: invalid memory reference)
> Error: Workspace Unit Tests failed
> ```
> 
> And no debug information when trying to debug with `rust-gdb`. I could only hope it would pass CI.

Hmmm that's weird 🤔 just like the wgpu tests were failing in your original post.

I don't seem to be running into this issue myself and neither is the CI.

laggui commented:
The CI is failing on macos wgpu tests for reasons totally unrelated to this PR

```
  failures:
  
  ---- tests::jit::module_conv_transpose3d::tests::test_conv_transpose3d_simple_2 stdout ----
  thread 'tests::jit::module_conv_transpose3d::tests::test_conv_transpose3d_simple_2' panicked at crates/burn-wgpu/src/lib.rs:90:5:
  Tensors are not approx eq:
    => Position 0: 238452 != -1.0466474294662476 | difference 238453.04664742947 > tolerance 0.0010000000000000002
  
  
  failures:
      tests::jit::module_conv_transpose3d::tests::test_conv_transpose3d_simple_2
  
  test result: FAILED. 1630 passed; 1 failed; 8 ignored; 0 measured; 0 filtered out; finished in 26.05s
  ```
